x/shield/keeper/staking_purchase.go

```diff
@@ -0,0 +1,187 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"fmt"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/certikfoundation/shentu/x/shield/types"
+)
+
+func (k Keeper) GetGlobalStakeForShieldPool(ctx sdk.Context) (pool sdk.Int) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetGlobalStakeForShieldPoolKey())
+	if bz != nil {
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &pool)
+	}
+	return sdk.NewInt(0)
+}
+
+func (k Keeper) SetGlobalShieldStakingPool(ctx sdk.Context, pool sdk.Int) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(pool)
+	store.Set(types.GetGlobalStakeForShieldPoolKey(), bz)
+}
+
+func (k Keeper) GetOriginalStaking(ctx sdk.Context, purchaseID uint64) (amount sdk.Int) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetOriginalStakingKey(purchaseID))
+	if bz != nil {
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &amount)
+	}
+	return sdk.NewInt(0)
+}
+
+func (k Keeper) SetOriginalStaking(ctx sdk.Context, purchaseID uint64, amount sdk.Int) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(amount)
+	store.Set(types.GetOriginalStakingKey(purchaseID), bz)
+}
+
+func (k Keeper) GetStakeForShield(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress) (purchase types.ShieldStaking, found bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetStakeForShieldKey(poolID, purchaser))
+	if bz != nil {
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &purchase)
+		found = true
+	}
+	return
+}
+
+func (k Keeper) SetStakeForShield(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress, purchase types.ShieldStaking) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(purchase)
+	store.Set(types.GetStakeForShieldKey(poolID, purchaser), bz)
+}
+
+func (k Keeper) AddStaking(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress, purchaseID uint64, amount sdk.Int) error {
+	pool := k.GetGlobalStakeForShieldPool(ctx)
+	pool = pool.Add(amount)
+	k.SetGlobalShieldStakingPool(ctx, pool)
```

Comment:
Should this state be updated later with the other ones in case `SendCoinsFromAccountToModule` fails?

Reply:
I guess that should be the right way to go, but if anything fails mid transaction nothing will be committed.

Reply:
oh I see.  i guess it doesn’t matter then.

Reply:
but since we also use it in the expiration processing I'll update it accordingly